### PR TITLE
Gen2 apriltag

### PR DESCRIPTION
**Pull request by csaba-luxonis**
WIP. Currently it runs on full frame with ~6 fps, with image manip downscaled it could be 30+ fps but it does not support RAW8 yet (crashes).

### CODE REVIEW

File: include/depthai-shared/datatype/RawAprilTags.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    int hamming;
+
+    // A measure of the quality of the binary decoding process: the
+    // average difference between the intensity of a data bit versus
+    // the decision threshold. Higher numbers roughly indicate better
+    // decodes. This is a reasonable measure of detection accuracy
+    // only for very small tags-- not effective for larger tags (where
+    // we could have sampled anywhere within a bit cell and still
+    // gotten a good detection.)
+    float decision_margin;
```
Comment (themarpe) [inline include/depthai-shared/datatype/RawAprilTags.hpp:34]: decisionMargin

File: include/depthai-shared/datatype/RawAprilTags.hpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    // The center of the detection in image pixel coordinates.
+    Point2f c;
+
+    // The corners of the tag in image pixel coordinates. These always
+    // wrap counter-clock wise around the tag.
+    dai::Rect p;
```
Comment (themarpe) [inline include/depthai-shared/datatype/RawAprilTags.hpp:41]: Use full names here - eg: `c` -> `center`
Same for `p` abbreviation

File: include/depthai-shared/datatype/RawAprilTags.hpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include "RawBuffer.hpp"
+#include "depthai-shared/common/Point3f.hpp"
+#include "depthai-shared/common/Rect.hpp"
+
+namespace dai {
+
+/**
+ * AprilTags structure.
+ */
+struct AprilTags {
```
Comment (themarpe) [inline include/depthai-shared/datatype/RawAprilTags.hpp:17]: `struct AprilTag` IMO

**Review by themarpe — CHANGES_REQUESTED** [review]
Consolidate naming. 
eg: `RawAprilTags` -> `AprilTagData` <- should be the same, just the Raw prepended. 

Also `AprilTagConfigData` is too close to AprilTagConfig (config message) or AprilTagData (data  message). Pick something else or expose otherwise. 

(Maybe: AprilTagThresholds & AprilTagType, which are then used by RawAprilTagConfig as members separatelly)

Otherwise AprilTag for the name of the node and AprilTagProperties for its properties is :ok_hand:

File: include/depthai-shared/datatype/RawAprilTagConfig.hpp [hunk 4]
```diff
@@ -5,34 +5,20 @@
...
-    enum class AprilTagType : std::int32_t { TAG_36H11 = 0, TAG_25H9, TAG_16H5, TAG_CIR21H7, TAG_CIR49H12, TAG_CUST48H12, TAG_STAND41H12, TAG_STAND52H13 };
+/// AprilTagType configuration data structure
+struct AprilTagType {
+    enum class Type : std::int32_t { TAG_36H11 = 0, TAG_25H9, TAG_16H5, TAG_CIR21H7, TAG_CIR49H12, TAG_CUST48H12, TAG_STAND41H12, TAG_STAND52H13 };
 
-    AprilTagType type = AprilTagType::TAG_36H11;
+    Type t = Type::TAG_36H11;
 };
-NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(AprilTagConfigData, type);
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(AprilTagType, t);
```
Comment (themarpe) [inline include/depthai-shared/datatype/RawAprilTagConfig.hpp]: Can just be:
```
enum class AprilTagType : std::int32_t { TAG_36H11 = 0, TAG_25H9, TAG_16H5, TAG_CIR21H7, TAG_CIR49H12, TAG_CUST48H12, TAG_STAND41H12, TAG_STAND52H13 };
```

**Review by themarpe — CHANGES_REQUESTED** [review]

File: include/depthai-shared/datatype/RawAprilTags.hpp (hunk 3 above)
Comment (themarpe) [reply include/depthai-shared/datatype/RawAprilTags.hpp:16]: @szabi-luxonis 
Address this issue as well

File: include/depthai-shared/datatype/DatatypeEnum.hpp [hunk 5]
```diff
@@ -16,6 +16,8 @@ enum class DatatypeEnum : std::int32_t {
     SpatialLocationCalculatorConfig,
     SpatialLocationCalculatorData,
     EdgeDetectorConfig,
+    AprilTagConfig,
+    AprilTagData,
```
Comment (themarpe) [inline include/depthai-shared/datatype/DatatypeEnum.hpp:20]: `AprilTags`

File: src/datatype/DatatypeEnum.cpp [hunk 6]
```diff
@@ -25,6 +25,8 @@ const std::unordered_map<DatatypeEnum, std::vector<DatatypeEnum>> hierarchy = {
          DatatypeEnum::StereoDepthConfig,
          DatatypeEnum::FeatureTrackerConfig,
          DatatypeEnum::TrackedFeatures,
+         DatatypeEnum::AprilTagConfig,
+         DatatypeEnum::AprilTagData,
```
Comment (themarpe) [inline src/datatype/DatatypeEnum.cpp:29]: AprilTags

File: src/datatype/DatatypeEnum.cpp [hunk 7]
```diff
@@ -41,6 +43,8 @@ const std::unordered_map<DatatypeEnum, std::vector<DatatypeEnum>> hierarchy = {
     {DatatypeEnum::StereoDepthConfig, {}},
     {DatatypeEnum::FeatureTrackerConfig, {}},
     {DatatypeEnum::TrackedFeatures, {}},
+    {DatatypeEnum::AprilTagConfig, {}},
+    {DatatypeEnum::AprilTagData, {}},
```
Comment (themarpe) [inline src/datatype/DatatypeEnum.cpp:47]: AprilTags

**Review by themarpe — COMMENTED** [review]
Please address the comments left as part of the response to the OP of the PR

File: include/depthai-shared/datatype/RawAprilTags.hpp (hunk 3 above)
Comment (szabi-luxonis) [reply include/depthai-shared/datatype/RawAprilTags.hpp:16]: AprilTag is the name of the node

Comment (themarpe) [reply include/depthai-shared/datatype/RawAprilTags.hpp:16]: Does it cause a conflict?

**Pull request merged**